IAR 8 feature branch merge

deepikabhavnani opened the pull request:
### Description

IAR 8.x feature branch merge.

@ARMmbed/mbed-os-test-team  - Please switch CI on master branch to IAR 8.32 after 5.11.5 release candidate is ready.

@ARMmbed/mbed-os-maintainers  - If PR https://github.com/ARMmbed/mbed-os/pull/9807 is merged before starting CI for this, then it should show up here. Else can update #9807 to be against master, if this is merged.

### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers
@SenRamakri 

### Release Notes

Updating the IAR toolchain version to IAR 8.32.

Commit:
Add Dynamic heap support to IAR

Two new block properties available in linker configuration files:
"expanding size" and "minimum size" for IAR 8.11.2 and above

Sample:
define block HEAP with expanding size, minimum size = 16K, alignment = 8 {};

This block will expand to consume all remaining available space in
the range where it is placed. If several such blocks end up in the same
range, they will share the remaining space.
Cannot place a block with expanding size inside another block with
expanding size, inside a block with maximum size, or inside an overlay.

Commit:
Update params in calls to LDREXW/STREXW to uint32_t

Commit:
Splited RAM section not compatible with dynamic HEAP.

RAM define/section if splited, IAR exanpding size considers just one
section while heap creation which results in smaller chunk for heap.
This commit is to unify the RAM section.

Commit:
Add Dynamic heap support to IAR

Two new block properties available in linker configuration files:
"expanding size" and "minimum size" for IAR 8.11.2 and above

Sample:
define block HEAP with expanding size, minimum size = 16K, alignment = 8 {};

This block will expand to consume all remaining available space in
the range where it is placed. If several such blocks end up in the same
range, they will share the remaining space.
Cannot place a block with expanding size inside another block with
expanding size, inside a block with maximum size, or inside an overlay.

Commit:
Update IAR project template as per IAR 8.32

Reference template file is used from IAR8.32 installation available at path
"Embedded Workbench 8.2\arm\config\template\ConvertToIAR\template_project.ewp"

Also added trustzone configs in iar_definitions.json

Commit:
Remove IAR supported from targets failing IAR 8.32 build

Few targets have libraries compatible to IAR 7.x, which fail with
IAR 8.x, IAR support for them is removed till updated library is
added.

CI failure seen if IAR is not added as supported target, trying with
release_version as empty string

List of targets affected:

1. UBLOX_EVK_ODIN_W2
2. MTB_UBLOX_ODIN_W2
3. MBED_CONNECT_ODIN
4. MTB_MXCHIP_EMW3166
5. MTB_ADV_WISE_1530
6. MTB_USI_WM_BN_BM_22
7. REALTEK_RTL8195AM

aqib-ublox committed:
adding updated binary for IAR v8.32

Commit:
Dynamic heap support will be breaking change for all targets in IAR 7.x
Reverting this feature till IAR 8.x is stable and supported in Mbed OS
for few releases

1. Revert "Add Dynamic heap support to IAR"

This reverts commit 8ee29bbfa2a4a1fd25a426a5d86391331f73e0ad.

2. Revert "Splited RAM section not compatible with dynamic HEAP."

This reverts commit a645bb2a6b0c4d82f2bf3db848f980f2fa4b1f7c.

3. Revert "Add Dynamic heap support to IAR"

This reverts commit 891e1f261b4729575f64225419acee163f23b6ae.

Commit:
Target_PSOC6 - Remove IAR supported from targets failing IAR 8.32 build

Commit:
Enable ODIN boards

Commit:
Set IAR version as 8.32 in tools

bridadan commented on tools/export/iar/ewp.tmpl:
```diff
@@ -1,974 +1,1043 @@
+                <option>
+                    <name>TrustZoneModes</name>
+                    <version>0</version>
+                    <state>{{trustZoneMode}}</state>
```
Any problem with making this variable match the name of this option?
```suggestion
                    <state>{{TrustZoneModes}}</state>
```

deepikabhavnani replied:
No problem in particular, I guess that choice was made since all other variables in python code had small initial char 
https://github.com/ARMmbed/mbed-os/pull/9808/files/66e93fd51a768cfe97c8d47e67f708f62ed8b178#diff-873b68ad68a5b46172dca8f9f82f5d13R142

bridadan replied:
AH true, I agree. Maybe just the plural then?
```suggestion
                    <state>{{trustZoneModes}}</state>
```

bridadan replied:
Ok yeah lets not change it for an `s`, it's Friday and my brain is gone 😆

bridadan reviewed (commented):
Looks ok, besides the disabled targets (though looks like you're addressing those)

Commit:
rand() by default does increment of 1, randomizing more

Random number generation in case of IAR 8, requires TLS support.
Thread local storage is not part of Mbed OS, test is updated to have random
numbers sparse, but in future random number creation should be moved to main
thread, or use some other logic for randomization instead of rand() call.

Commit:
Disable the cypress targets explicitly instead of base targets

Commit:
Merge pull request #9841 from deepikabhavnani/explicit_disable

Explicit disable Cypress targets

Commit:
Enhance targets.json with components

Adding storage components to boards

Commit:
Modifing general block device tests doe to targets enhancement

Due to targets enhancement some boards failed the general block device tests for flashiap component,
The fails were due to boards containing inconsistent sector sizes.
The tests were modified but should be improved to address the problem.
Rand() function issues were fixed.

0xc0170 commented on targets/TARGET_STM/TARGET_STM32F2/device/stm32f2xx_hal_def.h:
```diff
@@ -130,19 +130,19 @@ static inline  void atomic_set_u32(volatile uint32_t *ptr, uint32_t mask)
 	uint32_t newValue;
 	do {
-		newValue = (uint32_t)__LDREXW(ptr) | mask;
+		newValue = (uint32_t)__LDREXW((volatile uint32_t *)ptr) | mask;
```
@ARMmbed/team-st-mcd FYI

kjbracey replied:
I can't see what this change is doing. You're casting something to its own type.

deepikabhavnani replied:
This fix was added to feature branch more then a year ago. With updates to master and ST hal updates it is now irrelevant :-(
@0xc0170  - Shall i revert this as well?

Can see some fixes on similar terms on master https://github.com/ARMmbed/mbed-os/commit/c5a54382563b5b985833d2591ead8996cf427509

kjbracey replied:
Looks like a bad conflict resolution of that then - there was a `long` versus `uint32_t` mismatch, but that's gone now

Commit:
Add IAR8 binaries for USI_WM_BN_BM_22 and EMW3166 based chips

Commit:
Merge SPIF & QSPIF components test to general block device tests

The SPIF and QPIF components tests are fully merged into general block device tests and were deleted

theamirocohen commented on components/storage/blockdevice/COMPONENT_SPIF/TESTS/block_device/spif/main.cpp:
```diff
@@ -49,8 +49,9 @@ void basic_erase_program_read_test(SPIFBlockDevice &block_device, bd_size_t bloc
+    static unsigned block_seed = SPIF_TEST_NUM_OF_THREADS;
+    srand(block_seed);
+    block_seed += 2;
 
```
Please see #9855
SPIF and QSPIF tests were removed (merged fully into general block device tests)

deepikabhavnani replied:
@theamirocohen  We have created a Roll up PR to revert this fix and add storage changes to feature branch
https://github.com/ARMmbed/mbed-os/pull/9858 - Do we need to add #9855 as well to that?

theamirocohen replied:
Sounds good to me

Commit:
Revert "rand() by default does increment of 1, randomizing more"

This reverts commit f7a6d254f6702b64efada585f48b3be91ba1b836.

Commit:
Remove IAR from required toolchains list

Commit:
Enable Cypress for Mbed 5 release

Commit:
Revert "Update params in calls to LDREXW/STREXW to uint32_t"

This reverts commit 334b2d64b01724a467ea99210c752de39cc23b6e.

This commit is not required any more as changes for this are already part
of master branch

Commit:
Merge pull request #9858 from deepikabhavnani/roll_up_iar8

Roll up PR for feature-iar8

Commit:
Merge pull request #9863 from deepikabhavnani/old_commit_revert

Revert "Update params in calls to LDREXW/STREXW to uint32_t"

0xc0170 commented on targets/targets.json:
```diff
@@ -7327,12 +7335,12 @@
             "toolchains": ["ARM_STD", "GCC_ARM", "IAR"]
         },
-        "release_versions": ["5"],
+        "release_versions": [],
```
this should still have version 5 ?

cmonr replied:
If it does, I think it should come in via a followup PR, considering how much this PR is blocking.

0xc0170 replied:
just final check :) LGTM

0xc0170 reviewed (commented):
Just one line - might need an update

cmonr commented on tools/toolchains/iar.py:
```diff
@@ -33,7 +33,7 @@ class IAR(mbedToolchain):
     INDEX_PATTERN  = re.compile('(?P<col>\s*)\^')
     IAR_VERSION_RE = re.compile(b"IAR ANSI C/C\+\+ Compiler V(\d+\.\d+)")
-    IAR_VERSION = LooseVersion("7.80")
+    IAR_VERSION = LooseVersion("8.32")
```
🎉
